services/upgrader/upgrader_linux.go

```diff
@@ -27,48 +29,65 @@ import (
 	"github.com/mattermost/mattermost-server/v6/shared/mlog"
 )
 
-const mattermostBuildPublicKey = `-----BEGIN PGP PUBLIC KEY BLOCK-----
-
-mQENBFjZQxwBCAC6kNn3zDlq/aY83M9V7MHVPoK2jnZ3BfH7sA+ibQXsijCkPSR4
-5bCUJ9qVA4XKGK+cpO9vkolSNs10igCaaemaUZNB6ksu3gT737/SZcCAfRO+cLX7
-Q2la+jwTvu1YeT/M5xDZ1KHTFxsGskeIenz2rZHeuZwBl9qep34QszWtRX40eRts
-fl6WltLrepiExTp6NMZ50k+Em4JGM6CWBMo22ucy0jYjZXO5hEGb3o6NGiG+Dx2z
-b2J78LksCKGsSrn0F1rLJeA933bFL4g9ozv9asBlzmpgG77ESg6YE1N/Rh7WDzVA
-prIR0MuB5JjElASw5LDVxDV6RZsxEVQr7ETLABEBAAG0KU1hdHRlcm1vc3QgQnVp
-bGQgPGRldi1vcHNAbWF0dGVybW9zdC5jb20+iQFUBBMBCAA+AhsDBQsJCAcCBhUI
-CQoLAgQWAgMBAh4BAheAFiEEobMdRvDzoQsCzy1E+PLDF0R3SygFAmJOqWgFCQ03
-zUwACgkQ+PLDF0R3Syg/rQf8D5BgvVFnGuHDYNu2eiasZdfxmuhg1C7JGSLHqoCT
-SB/0SLLQyMeHsJLye/gbo3yhK8G9XYOm+obGF+NDxB0LtRaPv5Q6pIQYt88ZxOGA
-Kh6RG2DjYA5j410wYrN0mNzhudqnS2yZdyq215nEr7Z6l1T7L9OPcz0u0mF9RraQ
-nawzxbxc8mPuC5tMLTedViSkTYLgMY12TCSYhykseUIGrl/FBfMbmKwBHM52SZJh
-maBevuNymlFbODTciyE9Q7mJHkaamGKTXaa3Enlcf16oSoemawSBJuspaS0sZOW8
-dgi5l3V5YvfFvSk45axiZbnGYfN81G5mkSGAENSGSKVtMA==
-=kkvg
------END PGP PUBLIC KEY BLOCK-----`
-
-var upgradePercentage int64
-var upgradeError error
-var upgrading int32
+//go:embed pubkey.gpg
+var mattermostBuildPublicKeys []byte
+
+var (
+	upgradePercentage int64
+	m                 sync.Mutex
+	upgradeError      error
+	upgrading         int32
+)
 
 type writeCounter struct {
-	total  int64
-	readed int64
+	total int64
+	read  int64
 }
 
 func (wc *writeCounter) Write(p []byte) (int, error) {
 	n := len(p)
-	wc.readed += int64(n)
-	percentage := (wc.readed * 100) / wc.total
+	wc.read += int64(n)
+
+	if wc.total <= 0 {
+		// skip the percentage calculation for invalid totals
+		setUpgradePercentage(50)
+		return n, nil
+	}
+
+	percentage := (wc.read * 100) / wc.total
 	if percentage == 0 {
-		upgradePercentage = 1
-	} else if percentage == 100 {
-		upgradePercentage = 99
-	} else {
-		upgradePercentage = percentage
+		percentage = 1
+	} else if percentage >= 100 {
+		percentage = 99
 	}
+
+	setUpgradePercentage(percentage)
+
 	return n, nil
 }
 
+func getUpgradePercentage() int64 {
+	return atomic.LoadInt64(&upgradePercentage)
+}
+
+func setUpgradePercentage(to int64) {
+	atomic.StoreInt64(&upgradePercentage, to)
+}
+
+func getUpgradeError() error {
+	m.Lock()
+	defer m.Unlock()
+
+	return upgradeError
```

Comment:
Maybe we can use [atomic.Value](https://pkg.go.dev/sync/atomic#Value) for this?

Reply:
Unfortunately that's not possible. `atomic.Value` does not support storing `nil` [1] and after the first `Store` is fixed to this particular type [2], making it impossible to store different underlying error types (like we do).

[1] https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/sync/atomic/value.go;l=48
[2] https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/sync/atomic/value.go;l=77

Reply:
Ah interesting. Thanks!